codegeex-api.el

```diff
@@ -0,0 +1,88 @@
+;;; codegeex-api.el --- CodeGeeX For Emacs      -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2023  Samuel D
+
+;; Author: Samuel D <samueld@mailo.com>
+;; Keywords: codegeex, completion
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; This is handling the api of codegeex
+
+;;; Code:
+
+;; COMPLETION
+
+(defun codegeex-api--get-completion (prefix suffix lang callback)
+  "Invoke CodeGeeX completion API.
+
+This function will complete code between PREFIX and SUFFIX, which are usually
+the content before cursor and after cursor, and put the result to the current
+buffer. LANG is the programming lanuauge of the code
+CALLBACK is launched with the content of the buffer."
+  (let ((url (concat codegeex-endpoint "multilingual_code_generate_adapt"))
+        (data (codegeex-api--generate-json-data prefix suffix lang)))
+    (codegeex-api--url-retrieve url data callback)))
+
+(defun codegeex-api--url-retrieve (url json-data callback)
+  "Default url retrieve as POST with json data"
+  (let ((url-request-method "POST")
+        (url-request-extra-headers
+         '(("Content-Type" . "application/json")))
+        (url-request-data json-data))
+    (url-retrieve
+     url
+     (lambda (status init-buffer callback)
+       (let ((result (codegeex-api--get-json-result)))
+         (setq codegeex-response-cache result)
+         (with-current-buffer init-buffer
+           (funcall callback result))))
+     `(,(current-buffer) ,callback) t)))
```

Comment:
Will be the main api fetcher. With custom callback that take the json-result as argument. Can be used for the other api call that you have implemented by the way.